pkg/handlers/archive.go

```diff
@@ -0,0 +1,163 @@
+package handlers
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/mholt/archiver/v4"
+	log "github.com/sirupsen/logrus"
+)
+
+type ctxKey int
+
+const (
+	depthKey ctxKey = iota
+)
+
+var (
+	maxDepth = 5
+)
+
+// Archive is a handler for extracting and decompressing archives.
+type Archive struct {
+	maxSize int
+	size    int
+}
+
+// New sets a default maximum size and current size counter.
+func (d *Archive) New() {
+	d.maxSize = 20 * 1024 * 1024 // 20MB
+	d.size = 0
+}
+
+// FromFile extracts the files from an archive.
+func (d *Archive) FromFile(data io.Reader) chan ([]byte) {
+	ctx := context.Background()
+	archiveChan := make(chan ([]byte), 512)
+	go func() {
+		defer close(archiveChan)
+		err := d.openArchive(ctx, 0, data, archiveChan)
+		if err != nil {
+			if errors.Is(err, archiver.ErrNoMatch) {
+				return
+			}
+			log.WithError(err).Debug("Error unarchiving chunk.")
+		}
+	}()
+	return archiveChan
+}
+
+// openArchive takes a reader and extracts the contents up to the maximum depth.
+func (d *Archive) openArchive(ctx context.Context, depth int, reader io.Reader, archiveChan chan ([]byte)) error {
+	if depth >= maxDepth {
+		return fmt.Errorf("max archive depth reached")
+	}
+	format, reader, err := archiver.Identify("", reader)
+	if err != nil {
+		if errors.Is(err, archiver.ErrNoMatch) && depth > 0 {
```

Comment:
**Nit:** Negating this to make the if body smaller will make it easier to follow for me.